cmd/kubeconfig/create.go

```diff
@@ -89,54 +66,57 @@ Note: A certificate once signed cannot be revoked for a particular user`,
 			}
 			clusterAPIURL := nodeConfig.Spec.API.APIAddressURL()
 
-			caCert, err := os.ReadFile(path.Join(opts.K0sVars.CertRootDir, "ca.crt"))
-			if err != nil {
-				return fmt.Errorf("failed to read cluster ca certificate: %w, check if the control plane is initialized on this node", err)
-			}
-			caCertPath, caCertKey := path.Join(opts.K0sVars.CertRootDir, "ca.crt"), path.Join(opts.K0sVars.CertRootDir, "ca.key")
-			userReq := certificate.Request{
-				Name:   username,
-				CN:     username,
-				O:      groups,
-				CACert: caCertPath,
-				CAKey:  caCertKey,
-			}
-			certManager := certificate.Manager{
-				K0sVars: opts.K0sVars,
-			}
-			userCert, err := certManager.EnsureCertificate(userReq, "root")
+			kubeconfig, err := createUserKubeconfig(opts.K0sVars, clusterAPIURL, username, groups)
 			if err != nil {
 				return err
 			}
 
-			data := struct {
-				CACert     string
-				ClientCert string
-				ClientKey  string
-				User       string
-				JoinURL    string
-			}{
-				CACert:     base64.StdEncoding.EncodeToString(caCert),
-				ClientCert: base64.StdEncoding.EncodeToString([]byte(userCert.Cert)),
-				ClientKey:  base64.StdEncoding.EncodeToString([]byte(userCert.Key)),
-				User:       username,
-				JoinURL:    clusterAPIURL,
-			}
-
-			var buf bytes.Buffer
-
-			err = userKubeconfigTemplate.Execute(&buf, &data)
-			if err != nil {
-				return err
-			}
-			_, err = cmd.OutOrStdout().Write(buf.Bytes())
-			if err != nil {
-				return err
-			}
-			return nil
+			_, err = cmd.OutOrStdout().Write(kubeconfig)
+			return err
 		},
 	}
+
+	cmd.Flags().AddFlagSet(config.FileInputFlag())
 	cmd.Flags().StringVar(&groups, "groups", "", "Specify groups")
 	cmd.PersistentFlags().AddFlagSet(config.GetPersistentFlagSet())
 	return cmd
 }
+
+func createUserKubeconfig(k0sVars *config.CfgVars, clusterAPIURL, username, groups string) ([]byte, error) {
+	userReq := certificate.Request{
+		Name:   username,
+		CN:     username,
+		O:      groups,
+		CACert: filepath.Join(k0sVars.CertRootDir, "ca.crt"),
+		CAKey:  filepath.Join(k0sVars.CertRootDir, "ca.key"),
+	}
+	certManager := certificate.Manager{
+		K0sVars: k0sVars,
+	}
+	userCert, err := certManager.EnsureCertificate(userReq, "root")
+	if err != nil {
+		return nil, fmt.Errorf("failed generate user certificate: %w, check if the control plane is initialized on this node", err)
+	}
+
+	const k0sContextName = "k0s"
+	kubeconfig := clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{k0sContextName: {
+			Server:               clusterAPIURL,
+			CertificateAuthority: userReq.CACert,
```

Comment:
hmm, doesn't this just make it a path to the file? I think we'd want to use `CertificateAuthorityData` instead

Reply:
ah, right.